### PR TITLE
User properties

### PR DESCRIPTION
**Pull request by millerdev**
Add support for new data sources schema elements that allow user properties to be specified and referenced with hashtag expressions (`#user/property`). It also removes magic/ambiguity from the data sources schema. Finally, all hard-coded references to `#case` hashtags have been removed since the data sources schema may define arbitrary hashtag names.

All changes here should be entirely backward-compatible with the existing data sources schema provided by HQ. Until HQ is updated to provide a new schema containing user properties, the only end-user visible change is c51fd69.

Spec: https://docs.google.com/document/d/1kPVwZketyWl6aRo1ZNCM3EngkrTsgmFbnBXAv-FThKU

Review commits individually 🐠 

@emord @orangejenny

### CODE REVIEW

File: src/util.js [hunk 1]
```diff
@@ -363,8 +363,13 @@ define([
...
 
-    that.isCaseReference = function (value) {
-        return value.startsWith('#case/');
+    that.getReferenceName = function (value) {
+        var ref = /^#([^\/]+)\//.exec(value);
+        if (!ref) {
+            return "Form Reference";
+        }
+        ref = ref[1][0].toUpperCase() + ref[1].substring(1).toLowerCase();
+        return ref + " Reference";
```
Comment (millerdev) [inline src/util.js]: This is used for analytics. It returns "Case Reference" and "Form Reference" (the same categories that have been used for hashtag analytics up to this point), and will also send a new category for each new hashtag namespace, for exmaple "User Reference" for `#user/...` hashtags. FYI @dimagi/product

**Comment by millerdev** [comment]
I'm leaning toward not doing QA on this since there are minimal user-facing changes, and I'm pretty confident that all changes to data sources schema are backward compatible. However, I'm probably biased in favor of my own work. @emord or @orangejenny do you have thoughts on that?

File: tests/form.js [hunk 2]
```diff
@@ -527,7 +527,61 @@ define([
...
+                "/data": null,
+                "/data/": null,
+                "/data/prop": "xpath",
+
+                "#case": null,
+                "#case/": "is",
+                "#case/prop": "has xpath",
+                /* TODO should these pass? (they don't)
+                "#case/ ": null,
+                "#case/+": null,
```
Comment (emord) [inline tests/form.js]: Agree `#case/+` should be `null`. I think `#case/: "is"` makes sense, why do you think it should be `null`?

File: tests/form.js [hunk 3]
```diff
@@ -527,7 +527,61 @@ define([
...
+            });
+
+            _.each({
+                "": null,
+                "#": null,
+
+                "#form": null,
+                /* TODO should these pass? (they don't)
+                "#form/": "is",
+                "#form/prop": "has",
```
Comment (emord) [inline tests/form.js]: From my memory. I think  this may be due to only checking valid hashtags against #case due to being interweaved with the data sources code, agreed with your  comment that they should pass

File: src/richText.js [hunk 4]
```diff
@@ -424,12 +422,12 @@ define([
      */
     function makeBubble(form, xpath) {
         function _parseXPath(xpath, form) {
-            if (CASE_REF_REGEX.test(xpath)) {
+            if (!FORM_REF_REGEX.test(xpath)) {
```
Comment (emord) [inline src/richText.js]: I think this means that #user and #case will be noted as the with the same classes and icons. Guessing that that was intentional, but curious what were the arguments against using a different UI for them?

File: src/form.js [hunk 5]
```diff
@@ -174,12 +174,16 @@ define([
...
                 oldHashtags = form.hashtagMap;
+            form.hashtagNamespaces = {form: true};
             if (form.richText) {
                 // TODO always load hashtags, even when rich text is disabled
                 form.hashtagMap = _.clone(vellum.datasources.getHashtagMap({}));
                 form.invertedHashtagMap = _.invert(form.hashtagMap);
                 form.hashtagTransformations = vellum.datasources.getHashtagTransforms({});
                 form.hasCaseHashtags = vellum.datasources.isReady();
+                _.each(form.hashtagTransformations, function (x, prefix) {
+                    form.hashtagNamespaces[/^#([^\/]+)/.exec(prefix)[1]] = true;
```
Comment (emord) [inline src/form.js]: This regex could be extracted out to a constant with a name. Also I think something like `#user+case` would match here, but we probably only want to match letters right?

File: src/datasources.js [hunk 6]
```diff
@@ -281,14 +286,18 @@ define([
                             info.hashtag = '#case/' + ref.subset;
                         }
                     }
-                    path = "instance('" + source.id + "')" + source.path +
-                           "[" + ref.key + " = " + path + "]";
+                    var keyPath = path;
+                    path = "instance('" + source.id + "')" + source.path;
+                    if (ref.subset_filter && ref.subset_key && ref.subset) {
```
Comment (emord) [inline src/datasources.js]: Do you think it would be useful to have subsets be a list, so that there could be multiple filters here?

File: tests/databrowser.js [hunk 7]
```diff
@@ -481,5 +481,119 @@ define([
...
+                });
+            });
+
+            it("should drag/drop user property", function (done) {
+                var form = util.loadXML(""),
+                    mug = util.addQuestion("DataBindOnly", "mug"),
+                    input = $("[name=property-calculateAttr]"),
+                    editor = input.ckeditor().editor,
+                    widget = util.getWidget('property-calculateAttr');
+                widget.input.promise.then(function () {
```
Comment (emord) [inline tests/databrowser.js]: I think current tests usually initialize the widget in the `before` call. Any reason you didn't do that here? I don't have strong opinions other than would be nice to be consistent

File: src/databrowser.js [hunk 8]
```diff
@@ -138,11 +138,14 @@ define([
...
         function handleDrop(node, info, target) {
             var widget = widgets.util.getWidget(target, vellum),
-                path = widget.mug.form.richText ? node.hashtag : node.xpath;
+                path = widget.mug.form.richText ? node.hashtag : node.xpath,
+                id;
             while (info) {
-                var id = widget.addInstanceRef({id: info.id, src: info.uri});
-                if (id !== info.id) {
-                    path = widget.mug.form.updateInstanceQuery(path, id, info.id);
+                if (info.id && info.uri) {
```
Comment (emord) [inline src/databrowser.js]: Commit message says that this fixes a bug. What was the bug and should it be tested?

File: tests/datasources.js [hunk 9]
```diff
@@ -38,21 +38,95 @@ define([
                 "@id": {},
                 name: {}
             }
+        }, {
+            id: "commcaresession",
```
Comment (millerdev) [inline tests/datasources.js]: b0baecfedf71b358e8ef8a52883e83ffc606fa34

File: src/datasources.js (hunk 6 above)
Comment (millerdev) [inline src/datasources.js]: I think it could make sense if we ever needed it, but I'm reluctant to do that now since we don't.

File: src/datasources.js [hunk 10]
```diff
@@ -300,7 +307,14 @@ define([
         }
         function getNodes(source, path, info) {
             var nodes = _.chain(source && source.structure)
-                .map(node(source, path, info))
+                .map(function (item, id) {
+                    if (item.merge) {
```
Comment (millerdev) [inline src/datasources.js]: The latter. `merge: true` was added because case properties and user properties have different base paths in the casedb (`/session/data` and `/session/context` respectively), but we don't want `data` and `context` to show up in the tree.

File: src/datasources.js [hunk 11]
```diff
@@ -357,7 +362,7 @@ define([
             _.each(nodes, function (node) {
                 if (node.hashtag && !node.index) {
                     hashtags.map[node.hashtag] = node.xpath;
-                    hashtags.transforms[node.hashtagPrefix] = function (prop) {
+                    hashtags.transforms[node.sourceInfo.hashtag + '/'] = function (prop) {
```
Comment (millerdev) [inline src/datasources.js]: Do you mean an assertion? `node.sourceInfo.hashtag` will only ever have a hashtag value that was sent in the data sources schema, which is by definition a hashtag prefix.

File: src/logic.js [hunk 12]
```diff
@@ -23,7 +23,7 @@ define([
             "hintItext",
         ],
         NO_SELF_REFERENCES = _.without(XPATH_REFERENCES, 'constraintAttr'),
-        CASE_REF_ID = "#case/";
+        EXTERNAL_REF = "#";
```
Comment (millerdev) [inline src/logic.js]: This constant (variable) is only ever used as a key in the `forward` and `reverse` mappings. It is not (nor is it intended to be) used to match the beginning of a hashtag or path. It's simply a key that should never collide with a mug ufid, and it's value could be anything that meets that criteria. I figured `#` was good enough for that, but happy to entertain other suggestions.

File: tests/form.js (hunk 3 above)
Comment (millerdev) [inline tests/form.js]: I added these tests to verify things that previously worked continued to work after refactoring and happened to uncover these surprises. I think changing/fixing that is out of scope for this PR.

File: src/richText.js (hunk 4 above)
Comment (millerdev) [inline src/richText.js]: That's correct. The spec shows the same icons for case and user properties. @dimagi/product might have more comments on this?

File: src/form.js (hunk 5 above)
Comment (millerdev) [inline src/form.js]: b0baecfedf71b358e8ef8a52883e83ffc606fa34

The input keys here are hashtag hashtag transformations, which are always populated from the hashtags supplied in the data sources schema (or in the short term, `#case/` guesses). I don't think there is any way a value like `#user+case` could get here unless it was specified in the schema, in which case we'd likely want it here.

File: tests/databrowser.js (hunk 7 above)
Comment (millerdev) [inline tests/databrowser.js]: I see some tests do that and [some](https://github.com/dimagi/Vellum/blob/b8714350b636b5a4a05fed9d7a5d8dd5c912f815/tests/databrowser.js#L119) do it this way. In this case I wanted to reset the editor since I'm dragging a different value into the same property widget in the previous test. Is there something I'm overlooking?

File: src/databrowser.js (hunk 8 above)
Comment (millerdev) [inline src/databrowser.js]: It's hard to explain, but because of the `merge: true` flag, there is now an object in the `info._parent` hierarchy that does not have these attributes.

I had been trying to write a test to make this fail before adding `merge: true` to the test schemas, but was unable to. Once I added it, existing tests failed, and this fixes them.

File: tests/form.js (hunk 2 above)
Comment (millerdev) [inline tests/form.js]: There's a space after `case/ `.

File: src/datasources.js (hunk 10 above)
Comment (emord) [inline src/datasources.js]: Ah. I see. so merge is for the UI layer, not the data layer? 

slightly related: May be useful to have better documentation for this builder function? I find myself having to really dig into the details whenver i look at it

File: src/logic.js (hunk 12 above)
Comment (emord) [inline src/logic.js]: Yeah. I think `#` is good enough for now, but i can see someone getting confused in the future. I think there's a comment in here about not splitting the case references as it gets rid of important inofrmation. This could be fixed if that is ever fixed

File: src/datasources.js (hunk 10 above)
Comment (millerdev) [inline src/datasources.js]: Correct, merge is for the UI. I'll think about how to improve the documentation. Do you want to see that in this PR? The schema transformation is complicated, and the docstring at the top of the file is supposed to explain how it's interpreted. I wonder if the schema itself could be simplified so the code was easier to understand?

Comment (emord) [inline src/datasources.js]: I wasn't expecting it here. Just thinking out loud about how this function is always a little difficult to get spun up on
